include/ginkgo/core/base/executor.hpp

```diff
@@ -1143,26 +1213,61 @@ class CudaExecutor : public detail::ExecutorBase<CudaExecutor>,
         return cusparse_handle_.get();
     }
 
+    /**
+     * Get the closest CPUs
+     *
+     * @return  the array of CPUs closest to this device
+     */
+    std::vector<int> get_closest_cpus() const
+    {
+        return this->cuda_exec_info_.closest_cpu_ids;
+    }
+
+    /**
+     * Gets the exec info struct
+     *
+     * @return  the exec_info struct
+     */
+    cuda_exec_info get_exec_info() const { return this->cuda_exec_info_; }
+
+    /**
+     * Bind the current executor thread to the list of cpus
+     *
+     * @param  ids  the logical ids of the cpus.
+     * @param  num_ids  the number of cpus to bind
+     */
+    void bind_to_cpus(const int *ids, const size_type num_ids) const;
+
+    /**
+     * Bind the current executor thread to the list of cores
+     *
+     * @param  ids  the logical ids of the cores.
+     * @param  num_ids  the number of cores to bind
+     */
+    void bind_to_cores(const int *ids, const size_type num_ids) const;
+
 protected:
     void set_gpu_property();
 
     void init_handles();
 
     CudaExecutor(int device_id, std::shared_ptr<Executor> master,
                  bool device_reset = false)
-        : EnableDeviceReset{device_reset},
-          device_id_(device_id),
-          master_(master),
-          num_warps_per_sm_(0),
-          num_multiprocessor_(0),
-          major_(0),
-          minor_(0),
-          warp_size_(0)
+        : EnableDeviceReset{device_reset}, master_(master)
     {
-        assert(device_id < max_devices && device_id >= 0);
+        this->cuda_exec_info_.device_id = device_id;
+        this->cuda_exec_info_.num_cores = 0;
+        this->cuda_exec_info_.num_work_groups_per_core = 0;
+        assert(this->cuda_exec_info_.device_id < max_devices &&
+               this->cuda_exec_info_.device_id >= 0);
+        this->CudaExecutor::populate_exec_info(get_machine_topology());
```

Comment:
I would move this assert to the `populate_exec_info` overload and have this information be a no-op, or do almost nothing (maybe only set the device id) when the module itself is not compiled. This way, you get the executor instantiation failure *only* when you actually compile the module, which makes this more manageable.

This is how I somewhat implemented DPC++ actually.

Reply:
It turned out that I dont actually need this assert because there is an if clause that checks the same thing.